### PR TITLE
FISH-651 FISH-666 FISH-667 FISH-868 FISH-1372 FISH-1504 FISH-1526 FISH-1527 MicroProfile 4.0 & 4.1

### PR DESCRIPTION
**Pull request by Pandrex247**
Revert of https://github.com/payara/Payara-Enterprise-Documentation/pull/166 and merge of https://github.com/payara/Payara-Enterprise-Documentation/pull/154

Main PR: https://github.com/payara/Payara-Enterprise/pull/433
TCK Runners PR: https://github.com/payara/MicroProfile-TCK-Runners/pull/147 and https://github.com/payara/MicroProfile-TCK-Runners/pull/152

### CODE REVIEW

File: docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc [hunk 1]
```diff
@@ -79,15 +75,13 @@ image:microprofile/fault-tolerance.png[Set Fault Tolerance Configuration]
 [[using-asadmin-commands]]
 === Using Asadmin Commands
 
-NOTE: The configuration no longer allows configuring user defined executor to run the fault tolerance execution. Instead, Fault Tolerance uses two dedicated executors that are well suited for the usage pattern typically experienced when executing Fault Tolerance semantics. It's possible to configure the maximum size of these pools to scale them up or down.
-
 ==== `set-fault-tolerance-configuration`
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc:78]:
```suggestion
[[set-fault-tolerance-configuration]]
==== `set-fault-tolerance-configuration`
```

File: docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc [hunk 2]
```diff
@@ -96,21 +90,23 @@ Provides a way to set the configuration of the fault tolerance service of the ta
 
 ===== Command Options
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc:91]:
```suggestion
[[set-fault-tolerance-configuration-options]]
===== Command Options
```

File: docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc [hunk 3]
```diff
@@ -121,14 +117,11 @@ Provides a way to set the configuration of the fault tolerance service of the ta
 |===
 
 
-
-
 ===== Example
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc:120]:
```suggestion
[[set-fault-tolerance-configuration-example]]
===== Example
```

File: docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc [hunk 4]
```diff
@@ -121,14 +117,11 @@ Provides a way to set the configuration of the fault tolerance service of the ta
...
 
 [source,Shell]
 ----
-asadmin> set-fault-tolerance-configuration --asyncmaxpoolsize=2000
---delaymaxpoolsize=30 --target instance1
+asadmin> set-fault-tolerance-configuration --managedexecutorservicename=java:comp/DefaultManagedExecutorService --target instance1
 ----
 
 ==== `get-fault-tolerance-configuration`
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc:127]:
```suggestion
[[get-fault-tolerance-configuration]]
==== `get-fault-tolerance-configuration`
```

File: docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc [hunk 5]
```diff
@@ -140,7 +133,7 @@ Returns the current configuration options for the Fault Tolerance service on the
 
 ===== Command Options
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc:134]:
```suggestion
[[get-fault-tolerance-configuration-options]]
===== Command Options
```

File: docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc [hunk 6]
```diff
@@ -159,5 +152,56 @@ Returns the current configuration options for the Fault Tolerance service on the
...
+----
+
+
+[[microprofile-4-upgrade]]
+== Upgrading from MicroProfile 3.x to 4.x
+
+MicroProfile 4.0 brings with it a number of changes to MicroProfile Fault Tolerance. There are two incompatible changes
+with the previous version of MicroProfile Fault Tolerance.
+
+=== Metric Names and Scopes Changed
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc:165]:
```suggestion
[[names-and-scopes-changed]]
=== Metric Names and Scopes Changed
```

File: docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc [hunk 7]
```diff
@@ -159,5 +152,56 @@ Returns the current configuration options for the Fault Tolerance service on the
...
+MicroProfile specifications. Note that this means:
+
+* Metrics are now exported under `/metrics` and `/metrics/base`, instead of `/metrics` and `/metrics/application` as in previous versions.
+* In the JSON format, when metrics are retrieved from `/metrics` they appear in the base object rather than the application object.
+* In the OpenMetrics format, the names are prefixed with `base_` instead of `application_`.
+
+There unfortunately is not a workaround for this change, users must simply ensure that any dashboard or application
+making use of these metrics must be updated to look at the new locations.
+
+==== Example
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc:180]:
```suggestion
[[names-and-scopes-changed-example]]
==== Example
```

File: docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc [hunk 8]
```diff
@@ -159,5 +152,56 @@ Returns the current configuration options for the Fault Tolerance service on the
...
+application:ft.<name>.timeout.callsTimedOut.total
+----
+
+New metric:
+[source,Shell]
+----
+base:ft.timeout.calls.total{method="<name>", timedOut="true"}
 ----
+
+=== Lifecycle of CircuitBreakers and Bulkheads Specified
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/faulttolerance.adoc:194]:
```suggestion
[[lifecycle-specified]]
=== Lifecycle of CircuitBreakers and Bulkheads Specified
```

File: docs/modules/ROOT/pages/documentation/microprofile/jwt.adoc [hunk 9]
```diff
@@ -146,7 +146,6 @@ This properties file should contain the key `accepted.issuer` with as value the
 
 === Namespaced claims configuration
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/jwt.adoc:147]:
```suggestion
[[namespaced-claims-configuration]]
=== Namespaced claims configuration
```

File: docs/modules/ROOT/pages/documentation/microprofile/jwt.adoc [hunk 10]
```diff
@@ -175,6 +174,29 @@ Then the JSON Parser search for namespace prefixed claims, remove the namespace
 ==== *custom.namespace*
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/jwt.adoc:174]:
```suggestion
[[custom-namespace]]
==== *custom.namespace*
```

File: docs/modules/ROOT/pages/documentation/microprofile/jwt.adoc [hunk 11]
```diff
@@ -175,6 +174,29 @@ Then the JSON Parser search for namespace prefixed claims, remove the namespace
...
 When the `custom.namespace` property is set, it will always take precedence over the default namespace and be used instead.
 
+=== Disabling Type Claim Verification
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/jwt.adoc:177]:
```suggestion
[[disabling-type-claim-verification]]
=== Disabling Type Claim Verification
```

File: docs/modules/ROOT/pages/documentation/microprofile/jwt.adoc [hunk 12]
```diff
@@ -175,6 +174,29 @@ Then the JSON Parser search for namespace prefixed claims, remove the namespace
...
+>
+> Use of this Header Parameter is OPTIONAL.
+
+For this reason, some third-party token issuers may generate tokens that are not compatible with the MicroProfile JWT specification. The Payara Platform allows to set this verification off, so you can use the `disable.type.verification` custom property and set its value to `true` to this effect. 
+
+NOTE: This property has to be defined in the `payara-mp-jwt.properties` configuration file described in the previous section.
+
+IMPORTANT: Keep in mind that tokens which are missing their type claim and are propagated to other services running on other Eclipse MicroProfile runtimes might be rejected, as the specification mandates the inclusion of the claim.
+
+=== Caching the Public Key
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/jwt.adoc:195]:
```suggestion
[[caching-public-key]]
=== Caching the Public Key
```

File: docs/modules/ROOT/pages/documentation/microprofile/jwt.adoc [hunk 13]
```diff
@@ -175,6 +174,29 @@ Then the JSON Parser search for namespace prefixed claims, remove the namespace
...
+
+By default, the public key retrieved by the `mp.jwt.verify.publickey.location` configuration property will be cached in memory for 5 minutes after being read from either a local file or a remote location. You can modify this _"time-to-live"_ which determines how long the key stays cached in memory in the case you are dealing with long-lived keys. This is done by setting the `publicKey.cache.ttl` custom property to the number of milliseconds after which the public key should be refreshed.
+
+NOTE: This property has to be defined in the `payara-mp-jwt.properties` configuration file described in the previous section.
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/jwt.adoc:199]:
```suggestion
NOTE: This property has to be defined in the `payara-mp-jwt.properties` configuration file described in the previous section.

```

A space between NOTE and the next header to fix the formatting

File: docs/modules/ROOT/pages/documentation/microprofile/jwt.adoc [hunk 14]
```diff
@@ -175,6 +174,29 @@ Then the JSON Parser search for namespace prefixed claims, remove the namespace
...
 == Activating JWT Authentication
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/jwt.adoc:200]:
```suggestion
[[activating-jwt]]
== Activating JWT Authentication
```

File: docs/modules/ROOT/pages/documentation/microprofile/metrics.adoc [hunk 15]
```diff
@@ -11,6 +11,42 @@ For a full overview of the API, review https://github.com/eclipse/microprofile-m
...
-{mpMetricsSpecUrl}[MicroProfile Metrics 1.1] was released in MicroProfile 1.3
+[[microprofile-4-upgrade]]
+== Upgrading from MicroProfile 3.x to 4.x
 
-IMPORTANT: Version {mpMetricsVersion} of the Metrics specification does not define the behaviour of implementations where multiple applications are deployed to the same container. This limitation will be addressed in a later version of the Metrics specification.
+MicroProfile 4.0 brings with it a number of changes to MicroProfile Metrics, among these changes are some which break
+previous behaviour. The majority of the breaking changes only affect those using the MicroProfile Metrics API
+programmatically (rather than using the annotations).
+
+=== All Metrics Now Reusable
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/metrics.adoc:21]:
```suggestion
[[upgrade-to-4-metrics-reusable]]
=== All Metrics Now Reusable
```

File: docs/modules/ROOT/pages/documentation/microprofile/metrics.adoc [hunk 16]
```diff
@@ -11,6 +11,42 @@ For a full overview of the API, review https://github.com/eclipse/microprofile-m
...
+In Metrics 3.0 all metrics are now considered reusable, and as such a number of methods related to this have been
+removed from the API.
+
+There isn't a workaround for this, you must simply make sure that you're not using any of the methods from the API and
+that you were not expecting any of your custom metrics to not be reusable.
+
+===  MetadataBuilder.withOptional* Methods Removed
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/metrics.adoc:28]:
```suggestion
[[upgrade-to-4-with-optional-removed]]
===  MetadataBuilder.withOptional* Methods Removed
```

File: docs/modules/ROOT/pages/documentation/microprofile/metrics.adoc [hunk 17]
```diff
@@ -11,6 +11,42 @@ For a full overview of the API, review https://github.com/eclipse/microprofile-m
...
+Only relevant to those using the MicroProfile Metrics API programmatically (rather than using the annotations), these
+methods have simply been removed.
+
+You should instead use the `MetadataBuilder.with*` methods, which all accept `null` as a "not present" value (with the
+exception of the `withName` method which must have a defined value).
+
+As an example, if you were using the `MetadataBuilder.withOptionalDisplayName(String displayName)` method, you must now
+use the `MetadataBuilder.withDisplayName(String displayName)` method.
+
+=== Metadata.getDescription() and Metadata.getUnit() Return Type Changed
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/metrics.adoc:39]:
```suggestion
[[upgrade-to-4-return-type-changed]]
=== Metadata.getDescription() and Metadata.getUnit() Return Type Changed
```

File: docs/modules/ROOT/pages/documentation/microprofile/metrics.adoc [hunk 18]
```diff
@@ -11,6 +11,42 @@ For a full overview of the API, review https://github.com/eclipse/microprofile-m
...
+
+Only relevant to those using the MicroProfile Metrics API programmatically (rather than using the annotations), these
+methods now return a `String` rather than an `Optional<String>`.
+
+Two new methods have been added to the API for those who wish to still get an `Optional<String>` return type:
+`Metadata.description()` and `Metadata.unit()`
+
+=== Timer.update(long duration, TimeUnit unit) Removed
```
Comment (OndroMih) [inline docs/modules/ROOT/pages/documentation/microprofile/metrics.adoc:47]:
```suggestion
[[upgrade-to-4-timer-methods-removed]]
=== Timer.update(long duration, TimeUnit unit) Removed
```

**Review by OndroMih — CHANGES_REQUESTED** [review]

File: docs/modules/ROOT/pages/documentation/microprofile/config/README.adoc [hunk 19]
```diff
@@ -647,3 +647,85 @@ asadmin> set-config-cache
     --duration=30
     --target=myAppCluster
 ----
+
+[[microprofile-4-upgrade]]
+== Upgrading from MicroProfile 3.x to 4.x
+
+MicroProfile 4.0 brings with it a number of changes to MicroProfile Config. There are three incompatible changes
+with the previous version of MicroProfile Config, though one is only important when writing your own _ConfigSources_.
```
Comment (MarkWareham) [inline docs/modules/ROOT/pages/documentation/microprofile/config/README.adoc:655]: This doesnt make alot of sense to me. Seems like a fragment of information.

**Review by MarkWareham — APPROVED** [review]
LGTM

**Pull request merged**